src/linker/Linker.Steps/ResolveFromAssemblyStep.cs

```diff
@@ -86,8 +86,10 @@ protected virtual void ProcessLibrary (AssemblyDefinition assembly)
 
 		public static void ProcessLibrary (LinkContext context, AssemblyDefinition assembly, RootVisibility rootVisibility = RootVisibility.Any)
 		{
-			var action = rootVisibility == RootVisibility.Any ? AssemblyAction.Copy : AssemblyAction.Link;
-			context.SetAction (assembly, action);
+			if (!context.Actions.Contains (assembly.Name.Name)) {
+				var action = rootVisibility == RootVisibility.Any ? AssemblyAction.Copy : AssemblyAction.Link;
+				context.SetAction (assembly, action);
```

Comment:
SetAction does more than just setting the action. I think better fix could be to leave this part untouched and modify SetAction to something like this

```c#
		public void SetAction (AssemblyDefinition assembly, AssemblyAction defaultAction)
		{
			RegisterAssembly (assembly);

			if (!_actions.TryGetValue (assembly.Name.Name, out AssemblyAction action))
				action = defaultAction;

			Annotations.SetAction (assembly, action);
		}
```

Reply:
Great suggestion, thanks! I had thought Context.SetAction was used elsewhere, but now I realize that we mostly call Annotations.SetAction, and *Context.SetAction* is used only in these two places.